File: src/frontend/src/types/workflow/task.ts
/**
 * LLM Guardrail configuration for validating task outputs using an LLM agent.
 * Uses CrewAI's OSS LLMGuardrail class.
 */
export interface LLMGuardrailConfig {
  /** Validation criteria description that the LLM will use to validate task output */
  description: string;
  /** LLM model to use for validation (e.g., 'databricks-claude-sonnet-4-5') */
  llm_model?: string;
}

export interface Task {
  id: string;
  name: string;
  description: string;
  expected_output: string;
  tools: string[];
  tool_configs?: Record<string, unknown>;  // User-specific tool configuration overrides
  agent_id: string | null;
  async_execution: boolean;
  markdown: boolean;
  context: string[];
  config: {
    cache_response: boolean;
    cache_ttl: number;
    retry_on_fail: boolean;
    max_retries: number;
    timeout: number | null;
    priority: number;
    error_handling: 'default' | 'retry' | 'ignore' | 'fail';
    output_file: string | null;
    output_json: string | null;
    output_pydantic: string | null;
    callback: string | null;
    callback_config?: Record<string, unknown> | null;
    human_input: boolean;
    markdown: boolean;
    condition?: string;
    guardrail?: string | null;  // Code-based guardrail (function name)
    llm_guardrail?: LLMGuardrailConfig | null;  // LLM-based guardrail configuration
  };
  created_at?: string;
  updated_at?: string;
  output?: string;
  callback?: string;
  converter_cls?: string;
  llm_guardrail?: LLMGuardrailConfig | null;  // Top-level for sync with config
}

export interface TaskFormData extends Omit<Task, 'config' | 'context'> {
  async_execution: boolean;
  markdown: boolean;
  output_json?: string;
  output_pydantic?: string;
  output_file?: string;
  human_input?: boolean;
  retry_on_fail?: boolean;
  max_retries?: number;
  timeout?: number | null;
  priority?: number;
  error_handling?: 'default' | 'ignore' | 'retry' | 'fail';
  cache_response?: boolean;
  cache_ttl?: number;
  tools: string[];
  context: string[];
  agent_id: string | null;
}

export interface Tool {
  id: string;
  name: string;
  description: string;
}

export interface TaskDependency {
  task_name: string;
  required_before: string[];
}

export interface ConnectionAssignment {
  agent_name: string;
  tasks: Array<{ task_name: string }>;
}

export interface ConnectionResponse {
  assignments: ConnectionAssignment[];
  dependencies: TaskDependency[];
}

export interface SavedTasksProps {
  refreshTrigger?: number;
}

export interface AdvancedConfig {
  async_execution: boolean;
  cache_response: boolean;
  cache_ttl: number;
  callback: string | null;
  callback_config?: Record<string, unknown> | null;
  context: string[];
  dependencies: string[];
  error_handling: string;
  human_input: boolean;
  markdown: boolean;
  max_retries: number;
  output_file: string | null;
  output_json: string | null;
  output_parser: string | null;
  output_pydantic: string | null;
  priority: number;
  retry_on_fail: boolean;
  timeout: number | null;
  condition?: string;
  guardrail?: string | null;  // Code-based guardrail (function name)
  llm_guardrail?: LLMGuardrailConfig | null;  // LLM-based guardrail configuration
}

export interface TaskAdvancedConfigProps {
  advancedConfig: AdvancedConfig;
  onConfigChange: (field: string, value: string | number | boolean | null | Record<string, unknown>) => void;
  availableTasks: Task[];
}

export interface TaskSelectionDialogProps {
  open: boolean;
  onClose: () => void;
  onTaskSelect: (tasks: Task[]) => void;
  tasks: Task[];
  onShowTaskForm: () => void;
  fetchTasks: () => Promise<void>;
  openInCreateMode?: boolean;
}

export interface TaskGenerationDialogProps {
  open: boolean;
  onClose: () => void;
  onTaskGenerated: (task: Task) => void;
  selectedModel?: string;
} 

File: src/backend/src/schemas/task.py
import json
from datetime import datetime
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ConditionConfig(BaseModel):
    """Schema for task condition configuration."""

    type: str
    parameters: Dict[str, Any] = Field(default_factory=dict)
    dependent_task: Optional[str] = None


class LLMGuardrailConfig(BaseModel):
    """Schema for LLM guardrail configuration.

    LLM guardrails use an AI model to validate task outputs against custom criteria.
    This is the OSS-compatible guardrail using crewai.tasks.llm_guardrail.LLMGuardrail.
    """

    description: str = Field(..., description="Validation criteria description")
    llm_model: Optional[str] = Field(
        default=None,
        description="Deprecated/ignored — the guardrail validates with the model "
        "the run uses (the chat-input selection), resolved at execution.",
    )


class TaskConfig(BaseModel):
    """Schema for task configuration settings."""

    cache_response: Optional[bool] = None
    cache_ttl: Optional[int] = None
    retry_on_fail: Optional[bool] = None
    guardrail_max_retries: Optional[int] = None
    timeout: Optional[int] = None
    priority: Optional[int] = None
    error_handling: Optional[str] = None
    output_file: Optional[str] = None
    output_json: Optional[str] = None
    output_pydantic: Optional[str] = None
    callback: Optional[str] = None
    callback_config: Optional[Dict[str, Any]] = None
    human_input: Optional[bool] = None
    condition: Optional[ConditionConfig] = None
    guardrail: Optional[str] = None  # Code-based guardrail (function name)
    llm_guardrail: Optional[LLMGuardrailConfig] = (
        None  # LLM-based guardrail configuration
    )
    markdown: Optional[bool] = None


# Shared properties
class TaskBase(BaseModel):
    """Base Pydantic model for Tasks with shared attributes."""

    name: str
    description: str
    agent_id: Optional[str] = None
    expected_output: str
    tools: List[str] = Field(default_factory=list)
    tool_configs: Optional[Dict[str, Dict[str, Any]]] = Field(
        default_factory=dict
    )  # Tool-specific config overrides
    async_execution: bool = False
    context: List[Union[str, str]] = Field(default_factory=list)
    config: TaskConfig = Field(default_factory=TaskConfig)
    output_json: Optional[str] = None
    output_pydantic: Optional[str] = None
    output_file: Optional[str] = None
    output: Optional[Any] = None  # Accept any type of output (dict or string)
    markdown: bool = False
    callback: Optional[str] = None
    callback_config: Optional[Dict[str, Any]] = None
    human_input: bool = False
    converter_cls: Optional[str] = None
    guardrail: Optional[str] = None  # Code-based guardrail (function name)
    llm_guardrail: Optional[LLMGuardrailConfig] = (
        None  # LLM-based guardrail configuration
    )


# Properties to receive on task creation
class TaskCreate(TaskBase):
    """Pydantic model for creating a task."""

    pass


# Properties to receive on task update
class TaskUpdate(BaseModel):
    """Pydantic model for updating a task, all fields optional."""

    name: Optional[str] = None
    description: Optional[str] = None
    agent_id: Optional[str] = None
    expected_output: Optional[str] = None
    tools: Optional[List[str]] = None
    tool_configs: Optional[Dict[str, Dict[str, Any]]] = (
        None  # Tool-specific config overrides
    )
    async_execution: Optional[bool] = None
    context: Optional[List[Union[str, str]]] = None
    config: Optional[TaskConfig] = None
    output_json: Optional[str] = None
    output_pydantic: Optional[str] = None
    output_file: Optional[str] = None
    output: Optional[Any] = None  # Accept any type of output (dict or string)
    markdown: Optional[bool] = None
    callback: Optional[str] = None
    callback_config: Optional[Dict[str, Any]] = None
    human_input: Optional[bool] = None
    converter_cls: Optional[str] = None
    guardrail: Optional[str] = None  # Code-based guardrail (function name)
    llm_guardrail: Optional[LLMGuardrailConfig] = (
        None  # LLM-based guardrail configuration
    )


# Properties shared by models stored in DB
class TaskInDBBase(TaskBase):
    """Base Pydantic model for tasks in the database, including id and timestamps."""

    id: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator("llm_guardrail", mode="before")
    @classmethod
    def parse_llm_guardrail(cls, v):
        """Parse llm_guardrail if it's a JSON string from database."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (json.JSONDecodeError, TypeError):
                return None
        return v


# Properties to return to client
class Task(TaskInDBBase):
    """Pydantic model for returning tasks to clients."""

    pass


# Backward compatibility alias
TaskResponse = Task
